grouper/audit.py

```diff
@@ -25,6 +25,48 @@ def user_is_auditor(username):
     return False
 
 
+def controllers_are_auditors(group):
+    """Return whether not all owners/managers in a group (and below) are auditors
+
+    This is used to ensure that all of the people who can control a group (owners, managers)
+    and all subgroups (all the way down the tree) have audit permissions.
+
+    Raises:
+        UserNotAuditor: If a user is found that violates the audit training policy, then this
+            exception is raised.
+
+    Returns:
+        bool: True if the tree is completely controlled by auditors, False if not.
+    """
+    graph = Graph()
+    checked, queue = set(), [group.name]
+    while queue:
+        cur_group = queue.pop()
+        if cur_group in checked:
+            continue
+        details = graph.get_group_details(cur_group)
+        for chk_user, info in details["users"].iteritems():
+            if chk_user in checked:
+                continue
+            # Only examine direct members of this group, because then the role is accurate.
+            if info["distance"] == 1:
+                if info["rolename"] == "member":
+                    continue
+                if user_is_auditor(chk_user):
+                    checked.add(chk_user)
+                else:
+                    raise UserNotAuditor(
```

Comment:
Would it be better to report all of the violations at once?  Conceivably there could be a large number of subgroups and it would be pesky to find out about them one by one.